fw/http.c

```diff
@@ -4807,156 +4778,278 @@ tfw_h2_append_predefined_body(TfwHttpResp *resp, unsigned int stream_id,
+		tmp_t_tz += next->truesize;
+	} else if (next && next->len != next_len) {
+		/* Some frags from current skb was moved to the next skb. */
+		BUG_ON(next->len < next_len);
+		__ADJUST_SKB_LEN_CHANGE(sk, next, next_len);
+
+		tfw_tcp_propagate_dseq(sk, skb);
+		ss_skb_copy_cb(next, skb, clear_mask);
+		tmp_t_tz += (long)next->truesize - (long)next_truesize;
+	} else {
+		tfw_tcp_propagate_dseq(sk, skb);
+	}
+
+	BUG_ON(tmp_t_tz < 0);
+	*t_tz = tmp_t_tz;
+
+	return r;
+
+#undef __ADJUST_SKB_LEN_CHANGE
+}
+
+static unsigned char
+tfw_h2_prepare_frame_flags(TfwStream *stream, TfwFrameType type, bool end)
+{
+	unsigned char flags;
+
+	switch (type) {
+	case HTTP2_HEADERS:
+	case HTTP2_CONTINUATION:
+		flags = stream->xmit.b_len ? 0 : HTTP2_F_END_STREAM;
+		flags |= end ? HTTP2_F_END_HEADERS : 0;
+		break;
+	case HTTP2_DATA:
+		flags = end ? HTTP2_F_END_STREAM : 0;
+		break;
+	default:
+		BUG();
+	};
+
+	return flags;
+}
+
+static unsigned int
+tfw_h2_calc_frame_length(struct sock *sk, struct sk_buff *skb, TfwH2Ctx *ctx,
+			 TfwStream *stream, TfwFrameType type,
+			 unsigned int limit, unsigned int len,
+			 unsigned int processed, unsigned int *nskbs)
+{
+	unsigned  char tls_type =  ss_skb_tls_type(skb);
+	unsigned short clear_flag = (type == HTTP2_DATA ?
+		SS_F_HTTT2_FRAME_DATA :  SS_F_HTTT2_FRAME_HEADERS);
+	unsigned short other_flag = (type == HTTP2_DATA ?
+		SS_F_HTTT2_FRAME_HEADERS : SS_F_HTTT2_FRAME_DATA);
+	unsigned int max_sz = min3(ctx->rsettings.max_frame_sz, limit, len);
+	unsigned int frame_sz = skb->len - FRAME_HEADER_SIZE - processed;
+	struct sk_buff *next = skb, *skb_tail = skb;
+
+	if (type == HTTP2_DATA) {
+		BUG_ON(ctx->rem_wnd <= 0 || stream->rem_wnd <= 0);
+		max_sz = min3(max_sz, (unsigned int)ctx->rem_wnd,
+			      (unsigned int)stream->rem_wnd);
+	}
+
+	while (!tcp_skb_is_last(sk, skb_tail)) {
+		next = skb_queue_next(&sk->sk_write_queue, skb_tail);
+
+		if (frame_sz + next->len > max_sz)
+			break;
+		/* Don't put different message types into the same frame. */
+		if (ss_skb_tls_type(next) != tls_type)
+			break;
+		/* Don't agregate skbs with different frame types. */
+		if (ss_skb_flags(next) & other_flag)
+			break;
+		ss_skb_clear_flag(next, clear_flag);
+		ss_skb_set_flags(next, SS_F_HTTP2_FRAME_PREPARED);
+		(*nskbs)++;
+		frame_sz += next->len;
+		skb_tail = next;
 	}
 
-	if (local_response)
+	return min(max_sz, frame_sz);
+}
+
+static int
+tfw_h2_make_frames(struct sock *sk, struct sk_buff *skb, TfwH2Ctx *ctx,
+		   TfwStream *stream, TfwFrameType type, unsigned long *len,
+		   unsigned int mss_now, unsigned int limit,
+		   unsigned int *processed, bool *end, unsigned int *t_tz,
+		   unsigned int *nskbs)
+{
+	int r = 0;
+	char *data;
+	unsigned char buf[FRAME_HEADER_SIZE];
+	const TfwStr frame_hdr_str = { .data = buf, .len = sizeof(buf)};
+	TfwMsgIter it = {
+		.skb = skb,
+		.skb_head = ((struct sk_buff *)&sk->sk_write_queue),
+		.frag = -1
+	};
+	TfwFrameHdr frame_hdr = {};
+
+	if (limit <= FRAME_HEADER_SIZE)
```

Comment:
We return 0 and after in `tfw_h2_sk_prepare_xmit` we simply set  `ss_skb_set_flags(skb, SS_F_HTTT2_FRAME_DATA_DONE);` that looks like skipping skb without framing when limit is small.

Reply:
Yes fixed in the last commit